lib/chat_transcript_service.rb

```diff
@@ -108,8 +128,31 @@ def generate_markdown
   private
 
   def messages
-    @messages ||= ChatMessage.includes(:user, :uploads).where(
+    @messages ||= ChatMessage.includes(:user, :uploads, :reactions).where(
       id: @message_ids, chat_channel_id: @channel.id
     ).order(:created_at)
   end
+
+  ##
+  # Queries reactions and returns them in this format
+  #
+  # emoji   |  usernames  |  chat_message_id
+  # ----------------------------------------
+  # +1      | foo,bar,baz | 102
+  # heart   | foo         | 102
+  # sob     | bar,baz     | 103
+  def reactions
+    @reactions ||= DB.query(<<~SQL, @messages.map(&:id))
```

Comment:
This isn't immediately clear to me but potentially how many messages can the `@messages` array be?

Reply:
There is no limit, _however_ the current highest amount of messages is 100 at a time from the archiving service:

https://github.com/discourse/discourse-chat/blob/ee9724e8af18dfcf6cdd1da0f1ff6cdaac5b228f/lib/chat_channel_archive_service.rb#L61-L81

Theoretically I guess someone could click tons of messages in the UI to create a quote for. Maybe I could introduce a limit on the number of messages that the transcript service will accept?

Reply:
`ARCHIVED_MESSAGES_PER_POST` is 100

Reply:
ahh OK so there is a upper limit. I just wanted to make sure of that. Btw, can you check if this query is using an index? I had a look at the model and aside from a unique index which I do not think this query will use, we do not have any other indexes on the foreign keys.

Reply:
Yeah I will take a look, probably wouldn't hurt to add an index to the `chat_message_id`

Reply:
Even after adding an index to `chat_message_id` on that table it is not used for this query:

```
                                            Table "public.chat_message_reactions"
     Column      |              Type              | Collation | Nullable |                      Default
-----------------+--------------------------------+-----------+----------+----------------------------------------------------
 id              | bigint                         |           | not null | nextval('chat_message_reactions_id_seq'::regclass)
 chat_message_id | integer                        |           |          |
 user_id         | integer                        |           |          |
 emoji           | character varying              |           |          |
 created_at      | timestamp(6) without time zone |           | not null |
 updated_at      | timestamp(6) without time zone |           | not null |
Indexes:
    "chat_message_reactions_pkey" PRIMARY KEY, btree (id)
    "chat_message_reactions_index" UNIQUE, btree (chat_message_id, user_id, emoji)
    "index_chat_message_reactions_on_chat_message_id" btree (chat_message_id)

discourse_development=# EXPLAIN ANALYZE SELECT emoji, STRING_AGG(DISTINCT users.username, ',') AS usernames, chat_message_id
discourse_development-# FROM chat_message_reactions
discourse_development-# INNER JOIN users on users.id = chat_message_reactions.user_id
discourse_development-# WHERE chat_message_id IN (8,9,10,11,12,13,14)
discourse_development-# GROUP BY emoji, chat_message_id
discourse_development-# ORDER BY chat_message_id, emoji
discourse_development-# ;
                                                                QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=153.01..192.04 rows=1735 width=47) (actual time=2.532..4.248 rows=1735 loops=1)
   Group Key: chat_message_reactions.chat_message_id, chat_message_reactions.emoji
   ->  Sort  (cost=153.01..157.34 rows=1735 width=30) (actual time=2.511..2.611 rows=1735 loops=1)
         Sort Key: chat_message_reactions.chat_message_id, chat_message_reactions.emoji
         Sort Method: quicksort  Memory: 181kB
         ->  Hash Join  (cost=6.21..59.66 rows=1735 width=30) (actual time=0.142..0.668 rows=1735 loops=1)
               Hash Cond: (chat_message_reactions.user_id = users.id)
               ->  Seq Scan on chat_message_reactions  (cost=0.00..48.53 rows=1735 width=19) (actual time=0.020..0.298 rows=1735 loops=1)
                     Filter: (chat_message_id = ANY ('{8,9,10,11,12,13,14}'::integer[]))
               ->  Hash  (cost=5.54..5.54 rows=54 width=19) (actual time=0.092..0.092 rows=54 loops=1)
                     Buckets: 1024  Batches: 1  Memory Usage: 11kB
                     ->  Seq Scan on users  (cost=0.00..5.54 rows=54 width=19) (actual time=0.012..0.050 rows=54 loops=1)
 Planning time: 2.488 ms
 Execution time: 4.540 ms
(14 rows)
```

However it doesn't seem to be too heavy of a query, that's 4ms which includes the chat message I made yesterday with 1700 reactions. Skipping adding the index for now as it seems to make no difference, here is the same query with no index:

```
                                                                QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=153.01..192.04 rows=1735 width=47) (actual time=5.274..6.972 rows=1735 loops=1)
   Group Key: chat_message_reactions.chat_message_id, chat_message_reactions.emoji
   ->  Sort  (cost=153.01..157.34 rows=1735 width=30) (actual time=5.264..5.318 rows=1735 loops=1)
         Sort Key: chat_message_reactions.chat_message_id, chat_message_reactions.emoji
         Sort Method: quicksort  Memory: 181kB
         ->  Hash Join  (cost=6.21..59.66 rows=1735 width=30) (actual time=0.153..2.303 rows=1735 loops=1)
               Hash Cond: (chat_message_reactions.user_id = users.id)
               ->  Seq Scan on chat_message_reactions  (cost=0.00..48.53 rows=1735 width=19) (actual time=0.018..0.953 rows=1735 loops=1)
                     Filter: (chat_message_id = ANY ('{8,9,10,11,12,13,14}'::integer[]))
               ->  Hash  (cost=5.54..5.54 rows=54 width=19) (actual time=0.082..0.082 rows=54 loops=1)
                     Buckets: 1024  Batches: 1  Memory Usage: 11kB
                     ->  Seq Scan on users  (cost=0.00..5.54 rows=54 width=19) (actual time=0.008..0.043 rows=54 loops=1)
 Planning time: 0.571 ms
 Execution time: 7.092 ms
(14 rows)
```

Reply:
Ooo I just a query on dev and realized the `chat_message_reactions_index` will actually be used in cases where the filter dataset is large since `chat_message_id` is the first column in the unique index. All good here :)